plugin/pkg/shoot/managedseed/admission.go

```diff
@@ -0,0 +1,174 @@
+)
+
+// Register registers a plugin.
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return New()
+	})
+}
+
+// ManagedSeed contains listers and and admission handler.
+type ManagedSeed struct {
+	*admission.Handler
+	shootLister          corelisters.ShootLister
+	seedManagementClient seedmanagementclientset.Interface
+	readyFunc            admission.ReadyFunc
+}
+
+var (
+	_ = admissioninitializer.WantsInternalCoreInformerFactory(&ManagedSeed{})
+	_ = admissioninitializer.WantsSeedManagementClientset(&ManagedSeed{})
+
+	readyFuncs = []admission.ReadyFunc{}
+)
+
+// New creates a new ManagedSeed admission plugin.
+func New() (*ManagedSeed, error) {
+	return &ManagedSeed{
+		Handler: admission.NewHandler(admission.Delete),
+	}, nil
+}
+
+// AssignReadyFunc assigns the ready function to the admission handler.
+func (v *ManagedSeed) AssignReadyFunc(f admission.ReadyFunc) {
+	v.readyFunc = f
+	v.SetReadyFunc(f)
+}
+
+// SetInternalCoreInformerFactory gets Lister from SharedInformerFactory.
+func (v *ManagedSeed) SetInternalCoreInformerFactory(f coreinformers.SharedInformerFactory) {
+	shootInformer := f.Core().InternalVersion().Shoots()
+	v.shootLister = shootInformer.Lister()
+
+	readyFuncs = append(readyFuncs, shootInformer.Informer().HasSynced)
+}
+
+// SetSeedManagementClientset gets the clientset from the Kubernetes client.
+func (v *ManagedSeed) SetSeedManagementClientset(c seedmanagementclientset.Interface) {
+	v.seedManagementClient = c
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (v *ManagedSeed) ValidateInitialization() error {
+	if v.shootLister == nil {
+		return errors.New("missing shoot lister")
+	}
+	if v.seedManagementClient == nil {
+		return errors.New("missing garden seedmanagement client")
+	}
+	return nil
+}
+
+var _ admission.ValidationInterface = &ManagedSeed{}
+
+// Validate validates if the Shoot can be deleted. If the
+func (v *ManagedSeed) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	// Wait until the caches have been synced
+	if v.readyFunc == nil {
+		v.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+	if !v.WaitForReady() {
+		return admission.NewForbidden(a, errors.New("not yet ready to handle request"))
+	}
+
+	// Ignore all kinds other than Shoot
+	if a.GetKind().GroupKind() != core.Kind("Shoot") {
+		return nil
+	}
+
+	// Ignore updates to status or other subresources
+	if a.GetSubresource() != "" {
+		return nil
+	}
+
+	switch {
+	case a.GetName() == "":
+		return v.validateDeleteCollection(ctx, a)
+	default:
+		return v.validateDelete(ctx, a)
+	}
+}
+
+func (v *ManagedSeed) validateDeleteCollection(ctx context.Context, a admission.Attributes) error {
+	shootList, err := v.shootLister.List(labels.Everything())
```

Comment:
I think, we should use a direct lookup here instead of a lister, which might be out of date, to not miss any shoots.

Reply:
By direct lookup you mean invoking the `List` method on a client, correct? BTW, I was "inspired" by https://github.com/gardener/gardener/blob/b1c8b23628d64f4dc3abb62ba33ca34afeafe7bf/plugin/pkg/shootstate/validator/admission.go#L144, if you are right there might be a bug there as well.

Reply:
> By direct lookup you mean invoking the List method on a client, correct?

Yes, exactly.

> if you are right there might be a bug there as well.

Yes, true.